### ISSUE TITLE
Feature Request: Support for Multiple Independent Metrics

### ISSUE THREAD

**Issue by cdelv** [google/flax#4906]
Hi,

I was looking into the metrics object for flax, and it's a very cool idea. However, it is hard to store multiple independent metrics. For example, if I want to store the loss and a score, I would like to do something like:

`metrics.update(loss=loss, score=score)`

The closest thing to this would be [nnx.metrics.MultiMetric](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/training/metrics.html#flax.nnx.metrics.MultiMetric), but this one can only store one metric; it just applies multiple operations to that same metric. 

Having a single metric object capable of logging everything would be very nice.

**Comment by vfdev-5** [comment]
Hi @cdelv, thanks for the feature request! To understand better your suggestion, can you please provide a code snippet of what would you like to have?

[nnx.metrics.MultiMetric](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/training/metrics.html#flax.nnx.metrics.MultiMetric) is intended to contain and update multiple metrics in a single call. Here is the example from the docs:
```python
from flax import nnx
import jax, jax.numpy as jnp

metrics = nnx.MultiMetric(
  accuracy=nnx.metrics.Accuracy(), loss=nnx.metrics.Average()
)

logits = jax.random.normal(jax.random.key(0), (5, 2))
labels = jnp.array([0, 1, 1, 1, 0])
logits2 = jax.random.normal(jax.random.key(1), (5, 2))
labels2 = jnp.array([0, 1, 1, 1, 1])

batch_loss = jnp.array([1, 2, 3, 4])
batch_loss2 = jnp.array([3, 2, 1, 0])

metrics.update(logits=logits, labels=labels, values=batch_loss)
metrics.compute()
{'accuracy': Array(0.6, dtype=float32), 'loss': Array(2.5, dtype=float32)}

metrics.update(logits=logits2, labels=labels2, values=batch_loss2)
metrics.compute()
{'accuracy': Array(0.4, dtype=float32), 'loss': Array(2., dtype=float32)}
```

The only difference with your code `metrics.update(loss=loss, score=score)` is that accuracy is computed from logits and labels inputs. In case if you would like to average the `score`, similarly to the `loss`, than you could do:
```python
from flax import nnx
import jax, jax.numpy as jnp

metrics = nnx.MultiMetric(
  score=nnx.metrics.Average(argname="batch_score"),
  loss=nnx.metrics.Average(argname="batch_loss"),
)

batch_loss = jnp.array([1, 2, 3, 4])
batch_score = jnp.array([0.5])

metrics.update(batch_score=batch_score, batch_loss=batch_loss)
print(metrics.compute())
# {'score': Array(0.5, dtype=float32), 'loss': Array(2.5, dtype=float32)}

batch_loss = jnp.array([2, 3, 4, 5])
batch_score = jnp.array([0.6])

metrics.update(batch_score=batch_score, batch_loss=batch_loss)
print(metrics.compute())
# {'score': Array(0.55, dtype=float32), 'loss': Array(3., dtype=float32)}
```


Please let me know what do you think and if I misunderstood your request. Thanks!

**Comment by cdelv** [comment]
Hi @vfdev-5,

Thank you for the response. It seems that nnx.Metrics already does what I was proposing. Following the example, I thought all the metrics stored had to come from the same data. The second example does what I want. 

Does nnx.Metrics can store a time series, for example, the training's loss history. Or should one transfer the data, store it somewhere, and reset it after every epoch?

Best regards

**Comment by vfdev-5** [comment]
@cdelv great that the example solves your problem.

> Does nnx.Metrics can store a time series, for example, the training's loss history. Or should one transfer the data, store it somewhere, and reset it after every epoch?

There is no built-in metric object to store a value history during one epoch. On one hand, one can use experiment tracking systems, e.g. TensorBoard to log these values and display the evolution on the plot. On the other hand, If we need to store values during the epoch and compute some complex aggregation (e.g. median of absolution error), than one can implement an `EpochMetric` as following: 
```python
class EpochMetric(nnx.Metric):
  def __init__(self, compute_fn: Callable):
    self.compute_fn = compute_fn

  def reset(self) -> None:
    self._predictions: List[Any] = []
    self._targets: List[Any] = []

  def update(self, *, y_preds, y_true) -> None:
    self._predictions.append(y_preds)
    self._targets.append(y_true)

  def compute(self):
    result = self.compute_fn(self._predictions, self._targets)
    return result
```

**Comment by vfdev-5** [comment]
@cdelv I close this ticket as solved, please feel free to reopen this one if something is still need to be addressed or open another one with another topic. Thanks!
